### DISCUSSION TITLE
Can we store cookies to avoid re-signing in?

### DISCUSSION THREAD

**Discussion by vlourme** [sbertix/Swiftagram#123] [Q&A] [answered]
Hi, I would like to know better how caching works.

Actually, I'm storing user credentials on device keychain and when app is opened, I re-authenticate using keychain, if it does not work, the app re-ask user for credentials.

But I would to know if I can do better using `Secret` instead of storing credentials. I assume Instagram have a refresh `access_token` endpoint?

Also note, my app is based on `BasicAuthenticator` example with `KeychainStorage`.

Thanks 👍

**Comment by sbertix** [comment, marked as answer]
Hey @vlourme,

There's no refresh `access_token`, there's actually no `access_token` at all: the Instagram Private API works differently than regular APIs. 
If you check the example you'll see the user is always ever prompted to log in the first time: once it's done, and the `Secret` stored on `KeychainStorage`, it's gonna retrieve that `Secret` and never even prompt the login again. You only ever need to log back in if endpoints fail and they require you to authenticate the user again. 
**You should not store the user's credentials**.

Once you're authenticated with `BasicAuthenticator`, next time just call `KeychainStroage<Secret>().all().first` and get the first stored `Secret` and use it directly (or store the logged in user identifier and then retrieve the appropriate `Secret` like I'm doing in the _Followers_ example). 
Again, please, **never actually store username and passwords**, and do not reauthenticated unless it's needed, otherwise the account might be flagged as spam. 

You should totally check the _Followers_ example though and see how `Secret` and `KeychainStorage` can be used together 😊

**Reply by vlourme** [reply]
Ah thank you for explaining KeychainStorage!

I saw those methods such as `.all()` but I didn't not though I could bypass using BasicAuthenticator by retrieving `first()` entry.
That's perfect, exactly what I need.

Thank you a lot :+1:
